src/gfx/main.cpp

```diff
@@ -840,18 +847,42 @@ auto Palette::begin() -> decltype(colors)::iterator {
 	// Skip the first slot if reserved for transparency
 	return colors.begin() + options.hasTransparentPixels;
 }
+
 auto Palette::end() -> decltype(colors)::iterator {
-	return std::find(begin(), colors.end(), UINT16_MAX);
+	// Since the palette may contain gaps, we must scan from the end.
+	for (auto it = colors.end(); ;) {
+		--it; // The array is not empty, so it's fine to decrement pre-emptively.
+		// If the iterator points at a non-empty element, then it is the end
+		// (being followed only by empty slots). Thus one past is is one past the end.
+		if (*it != UINT16_MAX) {
+			return ++it;
+		}
+		if (it == colors.begin()) { // Is it still safe to decrement?
+			return it; // No? Then the palette is empty.
+		}
+	}
```

Comment:
I had to think a bit about the iteration logic here. I believe this would be simpler, and also valid even if it's somehow called on an empty palette:
```suggestion
	// Since the palette may contain gaps, we must scan from the end.
	for (auto it = colors.end(); it != colors.begin(); --it) {
		// If the iterator points after a non-empty element, then it is the end
		// (being followed only by empty slots). Thus one past it is one past the end.
		if (*(it - 1) != UINT16_MAX) {
			return it;
		}
	}
	return colors.end(); // The palette is empty.
```

Reply:
Aside from the last line not being `return colors.begin();`, yes, it's correct.

Reply:
Even in the general case, wouldn't `it - 1` be at least as cheap as `--it` followed by `++it` undoing it?

Reply:
No, because it has to create a temporary, i.e. make a copy.

Reply:
If only something like` std::string`'s`find_last_not_of` could apply here.

Reply:
Oh, I guess `rbegin()` and `rend()` could be used with the usual `std::find` to scan the array backwards. :man_facepalming:

Reply:
Would we need to manually implement `rbegin`/`rend`/a `reverse_iterator`?

Reply:
No, but we can use them to quickly find the index of the last non-`UINT16_MAX` element <var>n</var>, and then `return colors.begin() + n;`.

Reply:
Wait, actually, `colors` is a `std::array` which *is* reverse-iterable. So I think this is correct:
```c++
return (std::find_if(colors.rbegin(), colors.rend(), [](uint16_t c) { return c != UINT16_MAX; }) + 1).base();
```
(Note the avoidance of an off-by-one error with `.base()`.)